### PR TITLE
convert Heading component to TypeScript

### LINKED ISSUE #1002: [UIDS] TS conversion - Heading
---
name: 'ui-design-system' issue
projects: "user-interviews/15"
about: An issue template for 'ui-design-system'
title: "[UIDS]"
assignees: ''

---

## Description

Convert the `Heading` component to Typescript.

https://docs.userinterviews.com/?path=/docs/components-heading--docs

## Acceptance Criteria

- `Heading.jsx` is converted to `Heading.tsx`
- `Heading.stories.jsx` is converted to `Heading.stories.tsx`
- `Heading.stories.tsx` uses the updated [CSF3 format](https://storybook.js.org/docs/react/api/csf). Example here:
- `HEADING_PROPS` can be deprecated (assuming it's not being consumed on RS)

### PR DESCRIPTION
**Pull request by jasonbasuil**
closes #1002 

converts `Heading` component to TypeScript

Reviewers:

Primary: Gabe, Jeff
Secondary: Kyle (on PTO)

### CODE REVIEW

File: src/EmptyState/EmptyState.tsx [hunk 1]
```diff
@@ -36,7 +36,15 @@ const EmptyState = ({
...
-        <Heading className="EmptyState__title" level={4} size="lg" textAlign="center">{title}</Heading>
+        <Heading
+          className="EmptyState__title"
+          level={4}
+          size="lg"
+          textAlign="center"
+          weight="bold"
+          >
+            {title}
+        </Heading>
```
Comment (jasonbasuil) [inline src/EmptyState/EmptyState.tsx:47]: It's curious that TS doesn't recognize the default values set in the component. So in this case `weight` wasn't explicitly written here in the Heading component even though the default value for it is defined as `"bold"` and is respected.  


Screenshot 2023-09-18 at 6 49 12 PM

File: src/Heading/Heading.stories.tsx [hunk 2]
```diff
@@ -5,18 +5,57 @@ import Heading from './Heading';
...
+  },
   title: 'Components/Heading',
 };
 
 export default meta;
 type Story = StoryObj<typeof Heading>;
 
+/** 
+  A default `Heading` as a level `2` with size `xxl` and weight `bold`
+*/
```
Comment (jasonbasuil) [inline src/Heading/Heading.stories.tsx:23]: I'm testing out a different way of writing docs for each story. Not sure if we'll end up adopting this broadly. It's nice to know that you can write docs directly in the stories file and you should remember to update it if you ever change a story, though it does make the file a bit more bloated. Will discuss further with the rest of the DS team.

Comment (jeffbax) [reply src/Heading/Heading.stories.tsx:23]: do these docs get parsed by storybook?

**Review by jeffbax — APPROVED** [review]
looks fine although I'm curious if there is any automated conversion from JS/proptypes to TS that we can just run component by component and just do a look over if it mangled anything

**Comment by jasonbasuil** [comment]
> looks fine although I'm curious if there is any automated conversion from JS/proptypes to TS that we can just run component by component and just do a look over if it mangled anything

Not that I know of, other than doing a quick copy paste into chatGPT and asking for it to spit out something and then double checking. I'd say the bulk of the work is updating the `.stories.tsx` files to accept the updated CSF3 format in Storybook 7.

File: src/Heading/Heading.stories.tsx (hunk 2 above)
Comment (jasonbasuil) [reply src/Heading/Heading.stories.tsx:23]: yup

Screenshot 2023-09-20 at 2 54 06 PM

**Comment by jeffbax** [comment]
> > looks fine although I'm curious if there is any automated conversion from JS/proptypes to TS that we can just run component by component and just do a look over if it mangled anything
> 
> Not that I know of, other than doing a quick copy paste into chatGPT and asking for it to spit out something and then double checking. I'd say the bulk of the work is updating the `.stories.tsx` files to accept the updated CSF3 format in Storybook 7.

Lyft built one a while ago though its archived: https://marketplace.visualstudio.com/items?itemName=mohsen1.react-javascript-to-typescript-transform-vscode

Also: https://github.com/mskelton/ratchet

might be worth a quick experiment to see how it does

**Comment by jasonbasuil** [comment]
> > > looks fine although I'm curious if there is any automated conversion from JS/proptypes to TS that we can just run component by component and just do a look over if it mangled anything
> > 
> > 
> > Not that I know of, other than doing a quick copy paste into chatGPT and asking for it to spit out something and then double checking. I'd say the bulk of the work is updating the `.stories.tsx` files to accept the updated CSF3 format in Storybook 7.
> 
> Lyft built one a while ago though its archived: https://marketplace.visualstudio.com/items?itemName=mohsen1.react-javascript-to-typescript-transform-vscode
> 
> Also: https://github.com/mskelton/ratchet
> 
> might be worth a quick experiment to see how it does

that's actually pretty slick. It's pretty close to how it's implemented. thanks for sharing!
Screenshot 2023-09-20 at 3 05 01 PM

**Pull request merged**